File: NeetCode/03-Stack/ValidParanthesis.cpp
#include <iostream>
#include <stack>
#include <string>
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(int i=0;i<s.size();i++){
            char c = s[i];
            if(c == '{' || c == '[' || c == '('){
                st.push(c);
            }
            else{
                if(st.empty()){
                    return false;
                }
                char top = st.top();
                st.pop();
                if(c == '}' && !(top == '{')){
                    return false;
                }
                else if(c == ')' && !(top == '(')){
                    return false;
                }
                else if(c == ']' && !(top == '[')){
                    return false;
                }
                

            }
           
        }
        return st.empty();
    }
};
int main{
    string s;
    cin >> s;
    Solution sol;
    cout << sol.isValid(s);
}

File: NeetCode/03-Stack/ValidParanthesis.java
import java.util.*;
class Solution {
    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c == '{' || c == '[' || c == '('){
                stack.push(c);
            }
            else{
                if(stack.isEmpty()){
                    return false;
                }
                char top = stack.pop();
                if(c == '}' && !(top == '{')){
                    return false;
                }
                else if(c == ')' && !(top == '(')){
                    return false;
                }
                else if(c == ']' && !(top == '[')){
                    return false;
                }
                

            }
           
        }
        return stack.isEmpty();
    }
    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);
        String s = scan.nextLine();
        System.out.println(isValid(s));
        scan.close();
    }
}